test/e2e/storage/vsphere/vsphere.go

```diff
@@ -30,25 +45,188 @@ type VSphere struct {
 	Connect(ctx, vs)
 	finder := find.NewFinder(vs.Client.Client, true)
 	return finder.Datacenter(ctx, datacenterPath)
 }
 
+// GetDatacenter returns the DataCenter Object for the given datacenterPath
+func (vs *VSphere) GetDatacenterFromObjectReference(ctx context.Context, dc object.Reference) *object.Datacenter {
+	Connect(ctx, vs)
+	return object.NewDatacenter(vs.Client.Client, dc.Reference())
+}
+
 // GetAllDatacenter returns all the DataCenter Objects
 func (vs *VSphere) GetAllDatacenter(ctx context.Context) ([]*object.Datacenter, error) {
 	Connect(ctx, vs)
 	finder := find.NewFinder(vs.Client.Client, true)
 	return finder.DatacenterList(ctx, "*")
 }
 
-// GetVMByUUID gets the VM object from the given vmUUID
+// GetVMByUUID gets the VM object Reference from the given vmUUID
 func (vs *VSphere) GetVMByUUID(ctx context.Context, vmUUID string, dc object.Reference) (object.Reference, error) {
 	Connect(ctx, vs)
-	datacenter := object.NewDatacenter(vs.Client.Client, dc.Reference())
+	datacenter := vs.GetDatacenterFromObjectReference(ctx, dc)
 	s := object.NewSearchIndex(vs.Client.Client)
 	vmUUID = strings.ToLower(strings.TrimSpace(vmUUID))
 	return s.FindByUuid(ctx, datacenter, vmUUID, true, nil)
 }
+
+// GetFolderByPath gets the Folder Object Reference from the given folder path
+// folderPath should be the full path to folder
+func (vs *VSphere) GetFolderByPath(ctx context.Context, dc object.Reference, folderPath string) (vmFolderMor types.ManagedObjectReference, err error) {
+	Connect(ctx, vs)
+	datacenter := object.NewDatacenter(vs.Client.Client, dc.Reference())
+	finder := find.NewFinder(datacenter.Client(), true)
+	finder.SetDatacenter(datacenter)
+	vmFolder, err := finder.Folder(ctx, folderPath)
+	if err != nil {
+		framework.Logf("Failed to get the folder reference for %s. err: %+v", folderPath, err)
+		return vmFolderMor, err
+	}
+	return vmFolder.Reference(), nil
+}
+
+func (vs *VSphere) CreateVolume(volumeOptions *VolumeOptions, dataCenterRef types.ManagedObjectReference) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	Connect(ctx, vs)
+	datacenter := object.NewDatacenter(vs.Client.Client, dataCenterRef)
+	var (
+		err                     error
+		directoryAlreadyPresent = false
+	)
+	if datacenter == nil {
+		err = fmt.Errorf("datacenter is nil")
+		return "", err
+	}
+	if volumeOptions == nil {
+		volumeOptions = &VolumeOptions{}
+	}
+	if volumeOptions.Datastore == "" {
+		volumeOptions.Datastore = vs.Config.DefaultDatastore
+	}
+	if volumeOptions.CapacityKB == 0 {
+		volumeOptions.CapacityKB = DefaultDiskCapacityKB
+	}
+	if volumeOptions.Name == "" {
+		volumeOptions.Name = "e2e-vmdk-" + strconv.FormatInt(time.Now().UnixNano(), 10)
+	}
+	if volumeOptions.DiskFormat == "" {
+		volumeOptions.DiskFormat = DefaultDiskFormat
+	}
+	if volumeOptions.SCSIControllerType == "" {
+		volumeOptions.SCSIControllerType = DefaultSCSIControllerType
+	}
```

Comment:
what do you think about a separate func to handle vol defaults? Might be leverage-able?

Reply:
yes this is good idea. Added new func to handle defaults

```
// initVolumeOptions function sets default values for volumeOptions parameters if not set
func (vs *VSphere) initVolumeOptions(volumeOptions *VolumeOptions) {
.
.
}
```